### ISSUE TITLE
Allow `#` to broadcast over scalars 

### ISSUE THREAD

**Issue by adrhill** [google/jaxtyping#8]
Currently, broadcasting functions can be annotated using `#`. From the [API docs](https://github.com/google/jaxtyping/blob/7b698ae2153fb3e949a8acdceeb0e71f19d1ec00/API.md#shape):
> * `* name`: zero or more variable-size axes, e.g. `f32["*batch c h w"]`
> * Append `#` to a dimension size to indicate that it can be that size or equal to one -- i.e. broadcasting is acceptable.

However functions annotated in this way don't support broadcasting over scalar inputs, even though JAX does. 

I would welcome if `f["n#"]` (or `f["*n#"]`) allowed broadcasting over scalar inputs instead of having to use verbose union types.

## MWE
<details>

JAX' `jnp.add` will happily broadcast all of these inputs:
```python
import jax.numpy as jnp
from jaxtyping import f, jaxtyped
from typeguard import typechecked

x_float = jnp.float32(0.1)
y_float = jnp.float32(0.2)

x_array = jnp.array([x_float])
y_array = jnp.array([y_float])

xs = jnp.array([0.1, 0.2])
ys = jnp.array([0.2, 0.3])
```
```python-repl
>>> jnp.add(xs, ys)
DeviceArray([0.3, 0.5], dtype=float32)

>>> jnp.add(x_array, y_array)
DeviceArray([0.3], dtype=float32)

>>> jnp.add(x_float, y_float)
DeviceArray(0.3, dtype=float32)

>>> jnp.add(xs, x_array)
DeviceArray([0.2, 0.3], dtype=float32)

>>> jnp.add(xs, y_float)
DeviceArray([0.3, 0.4], dtype=float32)

>>> jnp.add(x_array, x_float)
DeviceArray([0.2], dtype=float32)
```
However annotating a wrapper functions `myadd` for scalar/vector addition will throw an error on any operation involving scalar inputs:
```python
@jaxtyped
@typechecked
def myadd(x: f["n#"], y: f["n#"]) -> f["n#"]:
    return jnp.add(x, y)
```
```python-repl
>>> myadd(xs, ys)
DeviceArray([0.3, 0.5], dtype=float32)

>>> myadd(x_array, y_array)
DeviceArray([0.3], dtype=float32)

>>> myadd(x_float, y_float)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/jaxtyping/decorator.py", line 34, in wrapper
    return fn(*args, **kwargs)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 1032, in wrapper
    check_argument_types(memo)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 875, in check_argument_types
    raise TypeError(*exc.args) from None
TypeError: type of argument "x" must be jaxtyping.array_types.f['n#']; got jaxlib.xla_extension.DeviceArray instead

>>> myadd(xs, x_array)
DeviceArray([0.2, 0.3], dtype=float32)

>>> myadd(xs, y_float)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/jaxtyping/decorator.py", line 34, in wrapper
    return fn(*args, **kwargs)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 1032, in wrapper
    check_argument_types(memo)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 875, in check_argument_types
    raise TypeError(*exc.args) from None
TypeError: type of argument "y" must be jaxtyping.array_types.f['n#']; got jaxlib.xla_extension.DeviceArray instead

>>> myadd(x_array, y_float)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/jaxtyping/decorator.py", line 34, in wrapper
    return fn(*args, **kwargs)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 1032, in wrapper
    check_argument_types(memo)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 875, in check_argument_types
    raise TypeError(*exc.args) from None
TypeError: type of argument "y" must be jaxtyping.array_types.f['n#']; got jaxlib.xla_extension.DeviceArray instead
```

Adding `*` for zero-dims, we get `f["*n#"]`, which makes ` myadd2(x_float, y_float)` pass:
```python
@jaxtyped
@typechecked
def myadd2(x: f["*n#"], y: f["*n#"]) -> f["*n#"]:
    return jnp.add(x, y)
```
```python-repl
>>> myadd2(xs, ys)
DeviceArray([0.3, 0.5], dtype=float32)

>>> myadd2(x_array, y_array)
DeviceArray([0.3], dtype=float32)

>>> myadd2(x_float, y_float)
DeviceArray(0.3, dtype=float32)

>>> myadd2(xs, x_array)
DeviceArray([0.2, 0.3], dtype=float32)

>>> myadd2(xs, y_float)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/jaxtyping/decorator.py", line 34, in wrapper
    return fn(*args, **kwargs)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 1032, in wrapper
    check_argument_types(memo)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 875, in check_argument_types
    raise TypeError(*exc.args) from None
TypeError: type of argument "y" must be jaxtyping.array_types.f['*n#']; got jaxlib.xla_extension.DeviceArray instead

>>> myadd2(x_array, y_float)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/jaxtyping/decorator.py", line 34, in wrapper
    return fn(*args, **kwargs)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 1032, in wrapper
    check_argument_types(memo)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 875, in check_argument_types
    raise TypeError(*exc.args) from None
TypeError: type of argument "y" must be jaxtyping.array_types.f['*n#']; got jaxlib.xla_extension.DeviceArray instead
```

Using just `*` for zero-dims, `myadd2(x_float, y_float)` passes, but `myadd3(xs, x_array)` fails as we might expect :
```python
@jaxtyped
@typechecked
def myadd3(x: f["*n"], y: f["*n"]) -> f["*n"]:
    return jnp.add(x, y)
```
```python-repl
>>> myadd3(xs, ys)
DeviceArray([0.3, 0.5], dtype=float32)

>>> myadd3(x_array, y_array)
DeviceArray([0.3], dtype=float32)

>>> myadd3(x_float, y_float)
DeviceArray(0.3, dtype=float32)

>>> myadd3(xs, x_array)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/jaxtyping/decorator.py", line 34, in wrapper
    return fn(*args, **kwargs)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 1032, in wrapper
    check_argument_types(memo)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 875, in check_argument_types
    raise TypeError(*exc.args) from None
TypeError: type of argument "y" must be jaxtyping.array_types.f['*n']; got jaxlib.xla_extension.DeviceArray instead

>>> myadd3(xs, y_float)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/jaxtyping/decorator.py", line 34, in wrapper
    return fn(*args, **kwargs)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 1032, in wrapper
    check_argument_types(memo)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 875, in check_argument_types
    raise TypeError(*exc.args) from None
TypeError: type of argument "y" must be jaxtyping.array_types.f['*n']; got jaxlib.xla_extension.DeviceArray instead

>>> myadd3(x_array, y_float)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/jaxtyping/decorator.py", line 34, in wrapper
    return fn(*args, **kwargs)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 1032, in wrapper
    check_argument_types(memo)
  File "/Users/funks/Library/Caches/pypoetry/virtualenvs/jax-md-playground-UowHBS0s-py3.10/lib/python3.10/site-packages/typeguard/__init__.py", line 875, in check_argument_types
    raise TypeError(*exc.args) from None
TypeError: type of argument "y" must be jaxtyping.array_types.f['*n']; got jaxlib.xla_extension.DeviceArray instead
```

In this specific case, this can worked around with union types, however these quickly get very verbose:
```python
from typing import Union

@jaxtyped
@typechecked
def myadd4(x: Union[f[""], f["n#"]], y: Union[f[""], f["n#"]]) -> Union[f[""], f["n#"]]:
    return jnp.add(x, y)
```
```python-repl
>>> myadd4(xs, ys)
DeviceArray([0.3, 0.5], dtype=float32)

>>> myadd4(x_array, y_array)
DeviceArray([0.3], dtype=float32)

>>> myadd4(x_float, y_float)
DeviceArray(0.3, dtype=float32)

>>> myadd4(xs, x_array)
DeviceArray([0.2, 0.3], dtype=float32)

>>> myadd4(xs, y_float)
DeviceArray([0.3, 0.4], dtype=float32)

>>> myadd4(x_array, y_float)
DeviceArray([0.3], dtype=float32)
```
</details>

**Comment by patrick-kidger** [comment]
Hmm. I think the general case here is actually even trickier: for example `(3, 4)` and `(4,)` will broadcast together; i.e. it's not just scalars.

I think the desired behaviour here is to have `isinstance(x, f['*foo#'])` (but not `*foo` or `foo#`) store not just a single shape, but all shapes it has been matched against so far. Then at each stage check that the latest shape is broadcast-compatible with all the shapes currently in the store.

This should be doable. I'd welcome a PR on this if this is something you want/need in the short term.

**Comment by patrick-kidger** [comment]
This should now be resolved in version 0.1.0.

Note that with this release, for consistency the `#` has been moved to the start of the dimension name, i.e. `*#foo` or `#*foo`, rather than `*foo#`.

**Comment by adrhill** [comment]
Thanks! :)
